File: client/src/main/scala/stasis/client/model/EntityRef.scala
package stasis.client.model

import java.nio.file.Path
import java.nio.file.Paths

import io.github.sndnv.fsi.Schemes

sealed trait EntityRef {
  def key: String

  def mapFilesystem(f: Path => Path): EntityRef
  def mapLibrary(f: (String, String) => (String, String)): EntityRef

  def flatMap(f: EntityRef => EntityRef): EntityRef = f(this)

  def asFilesystem: EntityRef.Filesystem
  def asLibrary: EntityRef.Library
}

object EntityRef {
  def default(key: String): EntityRef = {
    val components = Schemes.extract(key)
    Option(components.component1()) match {
      case Some(scheme) => EntityRef.Library(scheme = scheme, path = components.component2())
      case None         => EntityRef.Filesystem(path = Paths.get(key))
    }
  }

  final case class Filesystem(path: Path) extends EntityRef {
    override lazy val key: String = path.toAbsolutePath.toString

    override def mapFilesystem(f: Path => Path): EntityRef =
      EntityRef.Filesystem(path = f(path))

    override def mapLibrary(f: (String, String) => (String, String)): EntityRef =
      this

    override def asFilesystem: Filesystem =
      this

    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
    override def asLibrary: Library =
      throw new IllegalArgumentException(s"Requested a library reference but [$key] found")
  }

  final case class Library(scheme: String, path: String) extends EntityRef {
    override lazy val key: String = s"$scheme${Schemes.Delimiter}$path"

    override def mapFilesystem(f: Path => Path): EntityRef =
      this

    override def mapLibrary(f: (String, String) => (String, String)): EntityRef =
      EntityRef.Library(f(scheme, path))

    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
    override def asFilesystem: Filesystem =
      throw new IllegalArgumentException(s"Requested a filesystem reference but [$key] found")

    override def asLibrary: Library =
      this
  }

  object Library {
    def apply(pair: (String, String)): Library =
      Library(scheme = pair._1, path = pair._2)
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/model/EntityRef.kt
package stasis.client_android.lib.model

import io.github.sndnv.fsi.Schemes
import java.nio.file.Path
import java.nio.file.Paths

sealed class EntityRef {
    abstract val key: String

    abstract fun mapFilesystem(f: (Path) -> Path): EntityRef
    abstract fun mapLibrary(f: (String, String) -> Pair<String, String>): EntityRef

    fun flatMap(f: (EntityRef) -> EntityRef): EntityRef = f(this)

    abstract fun asFilesystem(): Filesystem
    abstract fun asLibrary(): Library

    data class Filesystem(val path: Path) : EntityRef() {
        override val key: String by lazy { path.toAbsolutePath().toString() }

        override fun mapFilesystem(f: (Path) -> Path): EntityRef =
            Filesystem(path = f(path))

        override fun mapLibrary(f: (String, String) -> Pair<String, String>): EntityRef =
            this

        override fun asFilesystem(): Filesystem =
            this

        override fun asLibrary(): Library =
            throw IllegalArgumentException("Requested a library reference but [$key] found")
    }

    data class Library(val scheme: String, val path: String) : EntityRef() {
        override val key: String by lazy { "$scheme${Schemes.Delimiter}$path" }

        override fun mapFilesystem(f: (Path) -> Path): EntityRef =
            this

        override fun mapLibrary(f: (String, String) -> Pair<String, String>): EntityRef {
            val (mappedScheme, mappedPath) = f(scheme, path)
            return Library(scheme = mappedScheme, path = mappedPath)
        }

        override fun asFilesystem(): Filesystem =
            throw IllegalArgumentException("Requested a filesystem reference but [$key] found")

        override fun asLibrary(): Library =
            this
    }

    companion object {
        fun default(key: String): EntityRef {
            val (scheme, path) = Schemes.extract(key)
            return when (scheme) {
                null -> Filesystem(path = Paths.get(key))
                else -> Library(scheme = scheme, path = path)
            }
        }
    }
}
